cli/azd/internal/vsrpc/server_service.go

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package vsrpc
+
+import (
+	"context"
+	"net/http"
+	"os"
+)
+
+// ServerService is the RPC server for the '/ServerService/v1.0' endpoint.
+type ServerService struct {
+}
+
+func NewServerService() *ServerService {
+	return &ServerService{}
+}
+
+// InitializeAsync is the server implementation of:
+// ValueTask<bool> InitializeAsync(string rootPath, CancellationToken cancellationToken);
+func (s *ServerService) InitializeAsync(ctx context.Context, rootPath string) (bool, error) {
+	// TODO(ellismg): Would love if we didn't Chdir here and instead just handled the root path as state we managed
+	// so we could multi-instance, but this a fast way to get stuff off the ground.
+	err := os.Chdir(rootPath)
+
+	return err == nil, err
+}
+
+// StopAsync is the server implementation of:
+// ValueTask StopAsync(CancellationToken cancellationToken);
+func (s *ServerService) StopAsync(ctx context.Context) error {
+	// TODO(ellismg): Need to think about how shutdown works. For now it is probably best to just have the
+	// client terminate `azd` once they know all outstanding RPCs have completed instead of trying to do a graceful
+	// shutdown.
+	return nil
+}
```

Comment:
So the assumption is that after the service is opened a client must always call `InitializeAsync` first to `azd init` the server.  Then `StopAsync` would be to stop the previous long running operation?

When provides the Go context when the service method is invoked?  I would imagine that context is the one we would be watching to cancel when appropriate vs a special `Stop` method on the server/service.

Reply:
> So the assumption is that after the service is opened a client must always call InitializeAsync first to azd init the server. 

Correct - The implementation of the RPC here is not great today but the general flow is going to look like:

1. VS launches `azd vs-server`
2. VS parses the connection information that's printed to standard out and connects to the server.
3. VS calls `InitializeAsync` to provide us with the root of the project (this will be the root folder, where the `.sln` is stored, not the directory where the AppHost lives).  This is where we will expect to find `azure.yaml` (or will create it if it doesn't exist when the first CreateEnvironmentAsync RPC comes in.

> Then StopAsync would be to stop the previous long running operation?

Right - When VS wants to shut us down they will call `StopAsync`.  I suspect the implementation will need to either wait for any long running operations to complete and then shut down (it may call cancel on all the contexts to help the operations end "quicker") and then terminate.  Note there may not be any long running operations in flight when `CancelAsync` is called.

> When provides the Go context when the service method is invoked? I would imagine that context is the one we would be watching to cancel when appropriate vs a special Stop method on the server/service.

Right - I haven't pushed this through yet since most of `azd` is also cancellation unaware, but what I expect is that a responsibility of the `Handler` method that is returned by the `Handler{Action|Func}N` methods will create a a child context by using `WithCancel` and store that in a map somewhere.  The server will also listen for cancellation RPCs and when one comes in, fetch the cancel method from that map and invoke it, which will cause the Context that is passed into the individual handler method to become cancelled.  We'll observe that and return early.

Suspect cancellation of these RPCs is a Preview 4 feature - since it requires both work in the server interface and us to start observing cancellation of the context at "interesting" points in `azd`.

Reply:
FWIW, I checked in a small .NET program that I've been using to poke at the server for my own testing, and it has a sort of end to end test that runs through a flow and logs a bunch of information.  I will try to write some end to tests in go next.